sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java

```diff
@@ -64,31 +64,38 @@ public ByteKeyRange currentRestriction() {
 
   @Override
   public SplitResult<ByteKeyRange> trySplit(double fractionOfRemainder) {
-    // TODO(BEAM-8871): Add support for splitting off a fixed amount of work for this restriction
-    // instead of only supporting checkpointing.
-
-    // If we haven't done any work, we should return the original range we were processing
-    // as the checkpoint.
-    if (lastAttemptedKey == null) {
-      ByteKeyRange rval = range;
-      // We update our current range to an interval that contains no elements.
-      range = NO_KEYS;
-      return SplitResult.of(range, rval);
+    // No split on an empty range.
+    if (NO_KEYS.equals(range)) {
+      return null;
```

Comment:
Should this method then be `@Nullable`  and/or in the parent class?

Reply:
It should be in the parent class.

@robertwb was going to take a look at fixing up the docs and method signatures to ensure that null, None, nil were allowed and what they meant under certain scenarios.

Reply:
I can update  java `RestrictionTracker` within this PR if that's preferable.

Reply:
That works for me.

Reply:
I am ok with `null` but couldn't we have modelled this with something more type friendly like `UnsplittableResult`

Reply:
Since restriction trackers return `RestrictionT`, we could only do this if we updated the `trySplit` method to return `SplitResult<RestrictionT>` where `SplitResult` would be something like `Optional`.

That could be useful if we wanted to plumb the failure reason for the split back to the runner.

Reply:
The `trySplit` has returned `SplitResult<RestrictionT>` as result. If we want the runner understand the failure reason, we also need to create a set of well known literals. If just for logging purpose, we can also log the msg at SDK level.

Reply:
Semantically, what we want to return here is the empty split. It is important that the infrastructure be able to recognize and empty split (of type RestrictionT) so that it won't try to further split/reschedule it. One could create a new SplitResult type for Java, but I'm unconvinced it's worth its weight. (Certainly None/nil feel more appropriate for Python and Go respectively.)